libs/core/include/core/serializers/binary_interface.hpp

```diff
@@ -0,0 +1,201 @@
+
+namespace fetch {
+namespace serializers {
+namespace interfaces {
+class BinaryInterface
+{
+public:
+  BinaryInterface(MsgPackByteArrayBuffer &serializer, uint64_t size)
+    : serializer_{serializer}
+    , size_{std::move(size)}
+  {}
+
+  void Write(uint8_t const *arr, uint64_t partial_size)
+  {
+    pos_ += partial_size;
+    if (pos_ > size_)
+    {
+      throw SerializableException(
+          std::string("exceeded number of allocated elements in array serialization"));
+    }
+
+    serializer_.WriteBytes(arr, partial_size);
+  }
+
+  MsgPackByteArrayBuffer &serializer()
+  {
+    return serializer_;
+  }
+
+private:
+  MsgPackByteArrayBuffer &serializer_;
+  uint64_t                size_;
+  uint64_t                pos_{0};
+};
+
+template <uint8_t C8, uint8_t C16, uint8_t C32>
+class BinaryConstructorInterface
+{
+public:
+  enum
+  {
+    CODE8  = C8,
+    CODE16 = C16,
+    CODE32 = C32
+  };
+
+  BinaryConstructorInterface(MsgPackByteArrayBuffer &serializer)
+    : serializer_{serializer}
+  {}
+
+  BinaryInterface operator()(uint64_t count)
+  {
+    if (created_)
+    {
+      throw SerializableException(std::string("Constructor is one time use only."));
+    }
+
+    if (count < (1 << 8))
+    {
+      uint8_t opcode = static_cast<uint8_t>(CODE8);
+      serializer_.Allocate(sizeof(opcode));
+      serializer_.WriteBytes(&opcode, sizeof(opcode));
+
+      uint8_t size = static_cast<uint8_t>(count);
+      size         = platform::ToBigEndian(size);
+      serializer_.Allocate(sizeof(size));
+      serializer_.WriteBytes(&size, sizeof(size));
+    }
+    else if (count < (1 << 16))
+    {
+      uint8_t opcode = static_cast<uint8_t>(CODE16);
+      serializer_.Allocate(sizeof(opcode));
+      serializer_.WriteBytes(&opcode, sizeof(opcode));
+
+      uint16_t size = static_cast<uint16_t>(count);
+      size          = platform::ToBigEndian(size);
+      serializer_.Allocate(sizeof(size));
+      serializer_.WriteBytes(reinterpret_cast<uint8_t *>(&size), sizeof(size));
+    }
+    else if (count < (1ull << 32))
+    {
+      uint8_t opcode = static_cast<uint8_t>(CODE32);
+      serializer_.Allocate(sizeof(opcode));
+      serializer_.WriteBytes(&opcode, sizeof(opcode));
+
+      uint32_t size = static_cast<uint32_t>(count);
+      size          = platform::ToBigEndian(size);
+      serializer_.Allocate(sizeof(count));
+      serializer_.WriteBytes(reinterpret_cast<uint8_t *>(&count), sizeof(count));
+    }
+    else
+    {
+      throw SerializableException(
+          error::TYPE_ERROR,
+          std::string("Cannot create container type with more than 1 << 32 elements"));
+    }
+
+    created_ = true;
+    return BinaryInterface(serializer_, count);
+  }
+
+  MsgPackByteArrayBuffer &serializer()
+  {
+    return serializer_;
+  }
+
+private:
+  bool                    created_{false};
+  MsgPackByteArrayBuffer &serializer_;
+};
+
+class BinaryDeserializer
```

Comment:
Move to a source file?

Reply:
Sure.

Reply:
Actually, after reinstating the count serializer, this interface is templated and I would post-pone this until such a time where we have collected all serialisation in a single library.